### ISSUE TITLE
leak in af matmul or af tile

### ISSUE THREAD

**Issue by WilliamTambellini** [arrayfire/arrayfire#2157]
Hello
I m facing a strange leak with af 3.6.0 (at least afcuda) : this is the chart from the used memory returned by nvidia-smi :
nvidia-smi --query-compute-apps=used_memory --format=csv,noheader,nounits -lms 50
I m ofcourse running only 1 compute app.

![computeusedmemafmatmulleak](https://user-images.githubusercontent.com/109458/40131936-33dd1894-58f0-11e8-8863-3582439715d1.png)

- at the beginning of my program, the first dozen of forward passes, everything is alright. See AF does not use more than 1GB.
- then (from snapshot 130 on the chart) every calls to a tiled 3d af::matmul is allocating 500MB of memory that does NOT appear in af getMemInfo() and is NOT deleted by af. Just before the crash : af mem info reports:   AllocBytes=437113856 AllocBuffers=727 LockBytes=437113856 LockBuffers=727 BUT nvidia-smi reports more than 5GB of used compute mem.
- few frames later, my soft of course crashes with a OOM exception. 

I cannot reproduce yet with a simple af code but the faulty operation is this one : 

```c++
 // big is 37682 x 512 x 1     
 // small is 512 x 1 x 6
 af::array tiledbig = af::tile(big, 1, 1, 6);
 output = af::matmul(tiledbig, small);
```

These leaking 500MB per matmul are strangely the size of the tiled big array : 37k * 512 * 6 * 4.

TBC

**Comment by WilliamTambellini** [comment]
Just more hints about the bug : 
- everything is alright for the 1st dozen of loop in my nn : mvidia-smi reports a very stable 1.7 GB usage, AF allocBytes about 1GB (everything alright)
- the AF allocBuffer is reaching 1000
- something happens in AF, probably a GC : allocBytes drops at 400 MB, allocBuffers drops at 535, nvidia-smi strangely still reports 1.7 GB
- the 3d tiled matmul runs (as explained above)
- nvidia-smi reports 2.3 GB, these extra 500MB will never be recovered provoking a OOM eventually later :   

...
10 : AllocBytes=1080420352 AllocBuffers=998 LockBytes=353137664 LockBuffers=463
nvidia-smi before matmul: 1753 MiB
nvidia-smi after matmul: 1753 MiB
 AllocBytes=1080420352 AllocBuffers=998 LockBytes=353137664 LockBuffers=463

11 : AllocBytes=405065728 AllocBuffers=535 LockBytes=356037632 LockBuffers=475
nvidia-smi before matmul:1721 MiB
nvidia-smi after matmul: 2311 MiB
 AllocBytes=1023653888 AllocBuffers=537 LockBytes=356037632 LockBuffers=475

**Comment by 9prady9** [comment]
Is this still an issue with 3.6.1 release ? Can you please verify.

**Comment by 9prady9** [comment]
@WilliamTambellini There were some mem leaks that we fixed in 3.6.1 release. Can you please re-test against 3.6.1 release and reopen the issue if it is still a problem. Closing the issue for now.
